Userland/Libraries/LibAudio/QOALoader.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/Span.h>
+#include <AK/Types.h>
+#include <LibAudio/Loader.h>
+#include <LibAudio/QOATypes.h>
+#include <LibAudio/SampleFormats.h>
+
+namespace Audio {
+
+// Decoder for the Quite Okay Audio (QOA) format.
+// NOTE: The QOA format is not finalized yet and this decoder might not be fully spec-compliant as of 2023-02-02.
+//
+// https://github.com/phoboslab/qoa/blob/master/qoa.h
+class QOALoaderPlugin : public LoaderPlugin {
+public:
+    explicit QOALoaderPlugin(NonnullOwnPtr<AK::SeekableStream> stream);
+    virtual ~QOALoaderPlugin() override = default;
+
+    static Result<NonnullOwnPtr<QOALoaderPlugin>, LoaderError> create(StringView path);
+    static Result<NonnullOwnPtr<QOALoaderPlugin>, LoaderError> create(Bytes buffer);
+
+    virtual LoaderSamples get_more_samples(size_t max_samples_to_read_from_input = 128 * KiB) override;
+
+    virtual MaybeLoaderError reset() override;
+    virtual MaybeLoaderError seek(int sample_index) override;
+
+    virtual int loaded_samples() override { return static_cast<int>(m_loaded_samples); }
+    virtual int total_samples() override { return static_cast<int>(m_total_samples); }
+    virtual u32 sample_rate() override { return m_sample_rate; }
+    virtual u16 num_channels() override { return m_num_channels; }
+    virtual DeprecatedString format_name() override { return "Quite Okay Audio (.qoa)"; }
+    virtual PcmSampleFormat pcm_format() override { return PcmSampleFormat::Int16; }
+
+private:
+    enum class IsFirstFrame : bool {
+        Yes = true,
+        No = false,
+    };
+
+    MaybeLoaderError initialize();
+    MaybeLoaderError parse_header();
+
+    MaybeLoaderError load_one_frame(Span<Sample>& target, IsFirstFrame is_first_frame = IsFirstFrame::No);
+    // Updates predictor values in lms_state so the next slice can reuse the same state.
+    MaybeLoaderError read_one_slice(QOA::LMSState& lms_state, Span<i16>& samples);
+    static QOA::UnpackedSlice unpack_slice(QOA::PackedSlice packed_slice);
+
+    // QOA's division routine for scaling residuals before final quantization.
+    static ALWAYS_INLINE i16 qoa_divide(i16 value, i16 scale_factor);
```

Comment:
I don't know if `ALWAYS_INLINE` works like this and would it not be simpler to just make `qoa_divide` a local static function in the cpp file?

(Also maybe a similar thing can be done for `unpack_slice`)

Reply:
ALWAYS_INLINE will work even if it's defined in the TU, not the header (that's the whole point), and yes, unpack_slice is another performance-critical function (called >2200 times per second for realtime playback) that should have this.

Reply:
Testing has been inconclusive; it seems to be faster but not a lot. Since it's definitely not slower I will do the optimization.